### PR TITLE
Add readme

### PR DESCRIPTION
**Pull request by tyurek**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,18 @@
-# tss-ecdsa ![Build Status](https://github.com/novifinancial/tss-ecdsa/workflows/Rust%20CI/badge.svg)
+# tss-ecdsa 
+This repo is a work-in-progress implementation of the UC-Secure threshold ECDSA signing protocol described in https://eprint.iacr.org/2021/060.pdf
```
Comment (indomitableSwan) [inline README.md:2]: Please include the full paper citation as well as the link.

File: README.md [hunk 2]
```diff
@@ -1 +1,18 @@
...
+Specifically, we are targeting the three-round presigning protocol (with quadratic overhead for identifying faulty actors).
+
+This codebase is generally intended to be network-agnostic. Programs take messages as input and potentially output some outgoing messages in response. The relaying of these messages is assumed to happen externally. However, a proof-of-concept example of such networking code can be found in examples/network.
+
+## Current State
+### What's Implemented
+Currently, the KeyGen protocol (Figure 5), Auxinfo (Figure 6, minus the key refreshing), and 3 round Presign (Figure 7) are implemented, along with the requisite zero-knowledge proofs. These protocols make use of (where appropriate) an echo-broadcast subprotocol in order to enforce non-equivocation of message contents
```
Comment (indomitableSwan) [inline README.md:9]: Specify that you are referencing figures from the paper.

Comment (indomitableSwan) [inline README.md:9]: Add a bit of context for what each of these protocols do. "Auxinfo" is particularly confusing naming.

This section would be better if you rephrase to focus on the implemented functionalities, i.e., if you reference the paper only for additional clarity and as a pointer for readers who would like further detail.

Comment (indomitableSwan) [reply README.md:9]: Also missing period at the end of L9.

File: README.md (hunk 1 above)
Comment (indomitableSwan) [reply README.md:2]: Missing period at the end of L2.

**Review by indomitableSwan — CHANGES_REQUESTED** [review]
Some minor nits, but otherwise this looks good.

File: README.md [hunk 3]
```diff
@@ -1 +1,35 @@
-# tss-ecdsa ![Build Status](https://github.com/novifinancial/tss-ecdsa/workflows/Rust%20CI/badge.svg)
+# tss-ecdsa
+
+This repo is a work-in-progress implementation of the ECDSA signing protocol described in "UC Non-Interactive, Proactive, Threshold ECDSA with Identifiable Aborts" [CGGMP] which can be found at https://eprint.iacr.org/2021/060.pdf.
```
Comment (indomitableSwan) [inline README.md:3]: I really mean the full citation. Including author names, venue published, etc.

File: README.md [hunk 4]
```diff
@@ -1 +1,35 @@
...
+
+##  What's Implemented
+
+### Key Generation (Figure 5 of CGGMP)
+
+KeyGen generates a threshold signing key, shares of which are distributed to each node. Every node outputs a private key along with the public keys of all other nodes. This only needs to be run once for a given set of nodes. 
+
+### Auxinfo (CGGMP Figure 6, minus the key refreshing)
+
+Auxinfo generates the auxilary information (such as Paillier keys) needed in order to compute presignatures. In CGGMP, this is done in parallel with key refreshing, however this codebase currently only implements the generation of auxilary information. This is run after KeyGen and only needs to be run once.
```
Comment (indomitableSwan) [inline README.md:17]: Please be more specific as to what the Auxinfo contains.

File: README.md (hunk 3 above)
Comment (indomitableSwan) [reply README.md:3]: Include the year published in the shorthand reference name.

File: README.md (hunk 4 above)
Comment (indomitableSwan) [reply README.md:17]: Question: Is key refreshing implemented in this codebase?

File: README.md [hunk 5]
```diff
@@ -1,16 +1,33 @@
...
+
+### Three Round Pre-signing (Figure 7 of CGGMP)
+
+Presign is a protocol to calculate pre-signatures, which can be computed before the message to be signed is known. Once a pre-signature is computed, a threshold signature can be easily calculated in one round of interaction. This protocol must be run for every message which is to be signed.
+
+### Other
+
+KeyGen, Auxinfo, and Presign are the three protocols needed in order to do threshold signing. All of the zero-knowledge proofs that underpin these protocols have been implemented, as has a echo-broadcast protocol which is needed in order to enforce non-equivocation of message contents.
```
Comment (indomitableSwan) [inline README.md:25]: typo: "a echo-" --> "an echo-"

**Review by indomitableSwan — CHANGES_REQUESTED** [review]
Thanks for updating. Still a few minor comments left to be resolved, and a question.

**Comment by indomitableSwan** [comment]
Also -- can you fix so that the CI checks pass?

File: README.md (hunk 4 above)
Comment (tyurek) [reply README.md:17]: It is not

File: README.md [hunk 6]
```diff
@@ -1,28 +1,28 @@
...
-### Key Generation (Figure 5 of CGGMP)
+### Key Generation (Figure 5 of CGGMP20)
 
 KeyGen generates a threshold signing key, shares of which are distributed to each node. Every node outputs a private key along with the public keys of all other nodes. This only needs to be run once for a given set of nodes. 
 
-### Auxinfo (CGGMP Figure 6, minus the key refreshing)
+### Auxinfo (CGGMP20 Figure 6, minus the key refreshing)
 
-Auxinfo generates the auxilary information (such as Paillier keys) needed in order to compute presignatures. In CGGMP, this is done in parallel with key refreshing, however this codebase currently only implements the generation of auxilary information. This is run after KeyGen and only needs to be run once.
+Auxinfo generates the auxilary information (Paillier keys and ring-Pedersen parameters) needed in order to compute presignatures. In CGGMP, this is done in parallel with key refreshing, however this codebase currently only implements the generation of auxilary information. This is run after KeyGen and only needs to be run once.
```
Comment (indomitableSwan) [inline README.md:17]: CGGMP20

**Review by indomitableSwan — CHANGES_REQUESTED** [review]
Thanks for the changes; this is much improved. Almost there.

File: README.md [hunk 7]
```diff
@@ -1,28 +1,28 @@
 # tss-ecdsa
 
-This repo is a work-in-progress implementation of the ECDSA signing protocol described in "UC Non-Interactive, Proactive, Threshold ECDSA with Identifiable Aborts" [CGGMP] which can be found at https://eprint.iacr.org/2021/060.pdf.
+This repo is a work-in-progress implementation of the ECDSA signing protocol described in "UC Non-Interactive, Proactive, Threshold ECDSA with Identifiable Aborts" [CGGMP20] which can be found at https://eprint.iacr.org/2021/060.pdf.
```
[the remark this answers is not in the document]
Comment (tyurek) [reply README.md:3]: Can you give an example of what this would look like?

[the remark this answers is not in the document]
Comment (indomitableSwan) [reply README.md:3]: quickest, maybe non-awkward fix:

This repo is a work-in-progress implementation of Canetti et al.'s threshold ECDSA protocol described in

[CGGMP20] R. Canetti, R. Gennaro, S. Goldfeder, N. Makriyannis, and U. Peled. UC non-interactive, proactive,  threshold ECDSA with identifiable aborts. In  ACM CCS 2020, pp. 1769–1787. ACM Press, 2020.

For details, see the [paper](https://eprint.iacr.org/2021/060.pdf).

[the remark this answers is not in the document]
Comment (indomitableSwan) [reply README.md:3]: i.e. and just delete the reference section, if we're not going to have any other references.

**Review by indomitableSwan — APPROVED** [review]
LGTM. Thanks for updating.

**Pull request merged**